devito/grid.py

```diff
@@ -0,0 +1,50 @@
+from devito.tools import as_tuple
+from devito.dimension import x, y, z
+
+import numpy as np
+
+__all__ = ['Grid']
+
+
+class Grid(object):
+
+    """
+    A cartesian grid that encapsulates a physical domain over which
+    to discretize :class:`Function`s.
+
+    :param shape: Shape of the domain region in grid points.
+    :param extent: Physical extent of the domain in m; defaults to a
+                   unit box of extent 1m in all dimensions.
+    :param origin: Physical coordinate of the origin of the domain;
+                   defaults to 0. in all dimensions.
+    :param dtype: Default data type to be inherited by all Functions
+                  created from this :class:`Grid`.
+    """
+
+    def __init__(self, shape, extent=None, origin=None, dimensions=None,
+                 dtype=np.float32):
+        self.shape = as_tuple(shape)
+        self.extent = as_tuple(extent or tuple(1. for _ in shape))
+        self.origin = as_tuple(origin or tuple(0. for _ in shape))
+        self.dtype = dtype
+
+        # TODO: Raise proper exceptions and logging
+        assert(self.dim == len(self.origin) == len(self.extent) == len(self.spacing))
```

Comment:
I'm not sure how this couldn't be the case? as in, given how the various fields are computed

Reply:
All of them could be user-provided, and I'm just asserting against user-level idiocy (pebkac).